### DISCUSSION TITLE
tutorials/intro-hotchocolate/09-resolvers

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#96] [course feedback]
# tutorials/intro-hotchocolate/09-resolvers

Replace mock data with live data using data sources in your resolvers.

https://www.apollographql.com/tutorials/intro-hotchocolate/09-resolvers

**Comment by xmione** [comment]
>This code works when I use SpotifyService but when I use my own custom Web API service, it doesn't work. One, they are different in openapi version. This is using 3.0.0, I'm using 3.0.1. Don't know the difference. Two it's got the object "$ref": "#/components/schemas/Spotify.Object.FeaturedPlaylists" which uses Guid Id, while in my api service, I have my $ref": "#/components/schemas/Client" which uses int ID. Don't know if that is the one affecting it. Three my api service uses two parameters string, baseUrl and HttpClient httpClient while it uses httpClient only. I tried adding the baseUrl parameter in the Program.cs but still did not work:

```
builder.Services.AddHttpClient<AccSolService>(client =>
{
    client.BaseAddress = new Uri("https://localhost:7040/"); 
});
```

this is my Client class:

```
using ApiClient = AccSolWeb.Client;

namespace AccSol.React
{
    [GraphQLDescription("The Company's client.")]
    public class Client
    {
        [GraphQLDescription("The ID for the client.")]
        [ID]
        public int Id { get; set; }
        [GraphQLDescription("The client code.")]
        public string? Code { get; set; }
        [GraphQLDescription("The client's name.")]
        public string? Name { get; set; }

        public Client(int id, string code, string name) 
        {
            Id = id;
            Code = code;
            Name = name;
        }
        
        public Client(ApiClient apiClient) 
        {
            Id = apiClient.Id;
            Code = apiClient.Code;
            Name = apiClient.Name;
        }
    }
}
```

Query.cs:

```
using AccSolWeb;
using SpotifyWeb;

namespace AccSol.React;

public class Query
{
   [GraphQLDescription("Playlists hand-picked to be featured to all users.")]
    public async Task<List<Playlist>> FeaturedPlaylists(SpotifyService spotifyService)
    {
        var response = await spotifyService.GetFeaturedPlaylistsAsync();
        // var items = response.Playlists.Items;
        // var playlists = items.Select(item => new Playlist(item));
        // return playlists.ToList();
        var list = response.Playlists.Items.Select(item => new Playlist(item)).ToList();
        return list;
    }

    [GraphQLDescription("Client list")]
    public async Task<List<Client>> Client(AccSolService service)
    {
        var response = await service.GetAllAsync();
        var list = response.Select(item => new Client(item)).ToList();

        return list;
    }

}
```

```
query Client{
    client{
        id
        code
        name
    }
}
```
>{
  "errors": [
    {
      "message": "Unexpected Execution Error",
      "locations": [
        {
          "line": 2,
          "column": 5
        }
      ],
      "path": [
        "client"
      ],
      "extensions": {
        "message": "Unable to resolve service for type 'System.String' while attempting to activate 'AccSolWeb.AccSolService'.",
        "stackTrace": "   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.ThrowHelperUnableToResolveService(Type type, Type requiredBy)\r\n   at lambda_method52(Closure, IServiceProvider, Object[])\r\n   at Microsoft.Extensions.Http.DefaultTypedHttpClientFactory`1.CreateClient(HttpClient httpClient)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitDisposeCache(ServiceCallSite transientCallSite, RuntimeResolverContext context)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass2_0.<RealizeService>b__0(ServiceProviderEngineScope scope)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)\r\n   at HotChocolate.Execution.Processing.MiddlewareContext.Service[T]()\r\n   at lambda_method11(Closure, IResolverContext)\r\n   at HotChocolate.Types.Helpers.FieldMiddlewareCompiler.<>c__DisplayClass9_0.<<CreateResolverMiddleware>b__0>d.MoveNext()\r\n--- End of stack trace from previous location ---\r\n   at HotChocolate.Execution.Processing.Tasks.ResolverTask.ExecuteResolverPipelineAsync(CancellationToken cancellationToken)\r\n   at HotChocolate.Execution.Processing.Tasks.ResolverTask.TryExecuteAsync(CancellationToken cancellationToken)"
      }
    }
  ]
}

how do I resolve this?
Thanks in advance.

**Reply by mabuyo** [reply]
Hi there @xmione! I can try to troubleshoot with you! Do you have a repo I could test this out with?

First thing I did notice, in `Program.cs` did you register the service? In the previous lesson, we used `RegisterService`:

```cs
builder.Services.AddGraphQLServer()
      .AddQueryType<Query>()
      .RegisterService<SpotifyService>();
```

**Reply by xmione** [reply]
Hi mabuyo,

Yes, Already registered the service. You can clone the [repo here](https://github.com/xmione/AccSol.React.git). Apologies in advance if you find the project messy, just been toying around with it.
Thanks in advance!

**Reply by mabuyo** [reply]
Alright, I think it has something to do with the `baseUrl` for the HttpClient. You can either set it in your `swagger2.json` and re-generate the client using `nswag`, or you can tweak the baseAddress code in your `Program.cs`

Option A, in `swagger2.json`:

```json
  "servers": [
    {
      "url": "https://localhost:7040/"
    }
  ],
```

Option B, in `Program.cs` (your code was close!):

```cs
builder.Services.AddHttpClient<AccSolService>().ConfigureHttpClient((sp, client) =>
{
  client.BaseAddress = new Uri("https://localhost:7040/");
});
```

I couldn't get your `AccSol.API` running to get it working all the way, but I think this is the missing piece? At least, I got a new error :) Let me know if either of those methods work.

**Reply by xmione** [reply]
I tried Option B before, it did not work. I just tried the Option A now and still it gives me this error in Banana Cake Pop IDE:
{
  "errors": [
    {
      "message": "Unexpected Execution Error",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "companyClients"
      ],
      "extensions": {
        "message": "Unable to resolve service for type 'System.String' while attempting to activate 'AccSolWeb.AccSolService'.",
        "stackTrace": "   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.ThrowHelperUnableToResolveService(Type type, Type requiredBy)\r\n   at lambda_method52(Closure, IServiceProvider, Object[])\r\n   at Microsoft.Extensions.Http.DefaultTypedHttpClientFactory`1.CreateClient(HttpClient httpClient)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitDisposeCache(ServiceCallSite transientCallSite, RuntimeResolverContext context)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass2_0.<RealizeService>b__0(ServiceProviderEngineScope scope)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)\r\n   at HotChocolate.Execution.Processing.MiddlewareContext.Service[T]()\r\n   at lambda_method11(Closure, IResolverContext)\r\n   at HotChocolate.Types.Helpers.FieldMiddlewareCompiler.<>c__DisplayClass9_0.<<CreateResolverMiddleware>b__0>d.MoveNext()\r\n--- End of stack trace from previous location ---\r\n   at HotChocolate.Execution.Processing.Tasks.ResolverTask.ExecuteResolverPipelineAsync(CancellationToken cancellationToken)\r\n   at HotChocolate.Execution.Processing.Tasks.ResolverTask.TryExecuteAsync(CancellationToken cancellationToken)"
      }
    }
  ],
  "data": null
}
Api works only if you have the latest sql server db engine and default instance, i think :D. Just pushed my changes.

**Reply by mabuyo** [reply]
Hmm, for Option A, did you re-generate the client using nswag?

Another thing we can try is adding `[Service]` attribute in `Query.cs`: `[Service] AccSolService service`

I'll try to push up a branch later today that is working for me!

**Reply by xmione** [reply]
yeah, completed the tutorials and followed it to the tee but when I tried to add my own API, got stuck from there. Can you include a simple .net core web api project and use it in the sample? That would be great. Will be looking forward to that.
Thanks much!

**Reply by mabuyo** [reply]
Hmm, with the latest code you pushed to the branch, I ran:

```
dotnet nswag openapi2csclient \                       
  /input:Data/swagger2.json \
  /classname:AccSolService \
  /namespace:AccSolWeb \
  /output:Data/AccSolService.cs
```

Which changed this section in `AccSolService.cs`:

```
        public AccSolService(System.Net.Http.HttpClient httpClient)
        {
            BaseUrl = "https://localhost:7040/";
            _httpClient = httpClient;
        }
```

With no other changes, I re-started the server, ran the query and got this error back: 

```
"extensions": {
        "message": "The SSL connection could not be established, see inner exception.",
        "stackTrace": "   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync
```

Which points to something wrong with the HTTP connection to the API, https://localhost:7040/ (as you mentioned, I probably don't have the correct sql server installed, and that's outside the scope of the course).

Can you try those steps?

**Reply by xmione** [reply]
sure. [Here](https://drive.google.com/file/d/15AKzP_544jYZCtxDYNYGHoKWQeRHTPlB/view?usp=sharing), see how I did it.

**Reply by mabuyo** [reply]
Thanks for the video! I can't tell for sure, but it looks like the `swagger2.json` file didn't include this:

```
  "servers": [
    {
      "url": "https://localhost:7040/"
    }
  ],
```

_With_ that change, it works for me. Opened up a PR for clarity. https://github.com/xmione/AccSol.React/pull/1/files

If you run that branch, does it work?

Hopefully we can get to the bottom of this!

**Reply by xmione** [reply]
wow, it worked! Thanks!

**Reply by mabuyo** [reply]
I'm so glad to hear it! 😊

**Reply by xmione** [reply]
If you want to pass the baseUrl parameter to the service, modify the registration of the service:
builder.Services.AddTransient<AccSolService>(sp =>
{
    var httpClientFactory = sp.GetRequiredService<IHttpClientFactory>();
    var httpClient = httpClientFactory.CreateClient();
    return new AccSolService(_baseAPIUrl, httpClient);
});
This also works. I've updated the code.

**Comment by charles-dataforensics** [comment]
need to add the "await" to the line in the tutorial : var response = spotifyService.GetFeaturedPlaylistsAsync();
it is in the comment and whole query.cs  but not the line that can be copied from the text.

**Comment by sphaxn** [comment]
I followed the steps in the example and in the explorer response got 
```
{
  "errors": [
    {
      "message": "Unexpected Execution Error",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "featuredPlaylists"
      ]
    }
  ]
}
```
Researching further I found that in SpotifyService.cs GetFeaturedPlaylistsAsync() in the call to Spotify it was getting a response of 503 Service Unavailable.  In the SpotifyService constructor I noticed that the Spotify base url was being set to 
```
https://spotify-demo-api-fe224840a08c.herokuapp.com/v1
```
I replaced this value with the base request url from the Spotify Swagger page referred to in chapter 8, REST as a data source:
```
https://spotify-clone-api-demo.up.railway.app/v1
```
With this url I got a successful response with the expected playlists.
